### PR TITLE
Internal representation (IR) compiler and evaluator

### PR DESCRIPTION
**Pull request by devyn**
# Description

This PR adds an internal representation language to Nushell, offering an alternative evaluator based on simple instructions, stream-containing registers, and indexed control flow. The number of registers required is determined statically at compile-time, and the fixed size required is allocated upon entering the block.

Each instruction is associated with a span, which makes going backwards from IR instructions to source code very easy.

Motivations for IR:

1. **Performance.** By simplifying the evaluation path and making it more cache-friendly and branch predictor-friendly, code that does a lot of computation in Nushell itself can be sped up a decent bit. Because the IR is fairly easy to reason about, we can also implement optimization passes in the future to eliminate and simplify code.
2. **Correctness.** The instructions mostly have very simple and easily-specified behavior, so hopefully engine changes are a little bit easier to reason about, and they can be specified in a more formal way at some point. I have made an effort to document each of the instructions in the docs for the enum itself in a reasonably specific way. Some of the errors that would have happened during evaluation before are now moved to the compilation step instead, because they don't make sense to check during evaluation.
3. **As an intermediate target.** This is a good step for us to bring the [`new-nu-parser`](https://github.com/nushell/new-nu-parser) in at some point, as code generated from new AST can be directly compared to code generated from old AST. If the IR code is functionally equivalent, it will behave the exact same way.
4. **Debugging.** With a little bit more work, we can probably give control over advancing the virtual machine that `IrBlock`s run on to some sort of external driver, making things like breakpoints and single stepping possible. Tools like `view ir` and [`explore ir`](https://github.com/devyn/nu_plugin_explore_ir) make it easier than before to see what exactly is going on with your Nushell code.

The goal is to eventually replace the AST evaluator entirely, once we're sure it's working just as well. You can help dogfood this by running Nushell with `$env.NU_USE_IR` set to some value. The environment variable is checked when Nushell starts, so config runs with IR, or it can also be set on a line at the REPL to change it dynamically. It is also checked when running `do` in case within a script you want to just run a specific piece of code with or without IR.

# Example

```nushell
view ir { |data|
  mut sum = 0
  for n in $data {
    $sum += $n
  }
  $sum
}
```
  
```gas
# 3 registers, 19 instructions, 0 bytes of data
   0: load-literal           %0, int(0)
   1: store-variable         var 904, %0 # let
   2: drain                  %0
   3: drop                   %0
   4: load-variable          %1, var 903
   5: iterate                %0, %1, end 15 # for, label(1), from(14:)
   6: store-variable         var 905, %0
   7: load-variable          %0, var 904
   8: load-variable          %2, var 905
   9: binary-op              %0, Math(Plus), %2
  10: span                   %0
  11: store-variable         var 904, %0
  12: load-literal           %0, nothing
  13: drain                  %0
  14: jump                   5
  15: drop                   %0          # label(0), from(5:)
  16: drain                  %0
  17: load-variable          %0, var 904
  18: return                 %0
```

# Benchmarks

All benchmarks run on a base model Mac Mini M1.

## Iterative Fibonacci sequence

This is about as best case as possible, making use of the much faster control flow. Most code will not experience a speed improvement nearly this large.

```nushell
def fib [n: int] {
  mut a = 0
  mut b = 1
  for _ in 2..=$n {
    let c = $a + $b
    $a = $b
    $b = $c
  }
  $b
}
use std bench
bench { 0..50 | each { |n| fib $n } }
```

IR disabled:

```
╭───────┬─────────────────╮
│ mean  │ 1ms 924µs 665ns │
│ min   │ 1ms 700µs 83ns  │
│ max   │ 3ms 450µs 125ns │
│ std   │ 395µs 759ns     │
│ times │ [list 50 items] │
╰───────┴─────────────────╯
```

IR enabled:

```
╭───────┬─────────────────╮
│ mean  │ 452µs 820ns     │
│ min   │ 427µs 417ns     │
│ max   │ 540µs 167ns     │
│ std   │ 17µs 158ns      │
│ times │ [list 50 items] │
╰───────┴─────────────────╯
```

explore ir view

## [gradient_benchmark_no_check.nu](https://github.com/nushell/nu_scripts/blob/main/benchmarks/gradient_benchmark_no_check.nu)

IR disabled:

```
╭───┬──────────────────╮
│ 0 │ 27ms 929µs 958ns │
│ 1 │ 21ms 153µs 459ns │
│ 2 │ 18ms 639µs 666ns │
│ 3 │ 19ms 554µs 583ns │
│ 4 │ 13ms 383µs 375ns │
│ 5 │ 11ms 328µs 208ns │
│ 6 │  5ms 659µs 542ns │
╰───┴──────────────────╯
```

IR enabled:

```
╭───┬──────────────────╮
│ 0 │       22ms 662µs │
│ 1 │ 17ms 221µs 792ns │
│ 2 │ 14ms 786µs 708ns │
│ 3 │ 13ms 876µs 834ns │
│ 4 │  13ms 52µs 875ns │
│ 5 │ 11ms 269µs 666ns │
│ 6 │  6ms 942µs 500ns │
╰───┴──────────────────╯
```

## [random-bytes.nu](https://github.com/nushell/nu_scripts/blob/main/benchmarks/random-bytes.nu)

I got pretty random results out of this benchmark so I decided not to include it. Not clear why.

# User-Facing Changes
- IR compilation errors may appear even if the user isn't evaluating with IR.
- IR evaluation can be enabled by setting the `NU_USE_IR` environment variable to any value.
- New command `view ir` pretty-prints the IR for a block, and `view ir --json` can be piped into an external tool like [`explore ir`](https://github.com/devyn/nu_plugin_explore_ir).

# Tests + Formatting
All tests are passing with `NU_USE_IR=1`, and I've added some more eval tests to compare the results for some very core operations. I will probably want to add some more so we don't have to always check `NU_USE_IR=1 toolkit test --workspace` on a regular basis.

# After Submitting
- [ ] release notes
- [ ] further documentation of instructions?
- [ ] post-release: publish `nu_plugin_explore_ir`

### CODE REVIEW

File: crates/nu-cmd-lang/src/core_commands/const_.rs [hunk 1]
```diff
@@ -46,6 +46,9 @@ impl Command for Const {
         call: &Call,
         _input: PipelineData,
     ) -> Result<PipelineData, ShellError> {
+        // This is compiled specially by the IR compiler. The code here is never used in when
```
Comment (fdncred) [inline crates/nu-cmd-lang/src/core_commands/const_.rs:49]: seems like grammar was wrong? i wouldn't really care but this same statement is all over the place. 😄 
```suggestion
        // This is compiled specially by the IR compiler. The code here is never used when
```

Comment (devyn) [reply crates/nu-cmd-lang/src/core_commands/const_.rs:49]: You're right, I don't know how I missed that 😅 

I'll find and replace it

**Comment by devyn** [comment]
Here's the `cargo bench` comparison.

In general, it's worse. For most of these benchmarks, the compile and parse time is more than the runtime. Nothing here is surprising, and I think we need to change some of these benchmarks to either not include the parse (so don't use `evaluate_commands`) or set up some benchmarks here that are eval focused that don't measure any parsing.

```
load_standard_lib                                  [  10.8 ms ...  12.2 ms ]     +12.38%*
record_create_1                                    [  14.5 us ...  15.4 us ]      +6.47%*
record_create_10                                   [  27.3 us ...  29.5 us ]      +8.31%*
record_create_100                                  [ 153.6 us ... 163.3 us ]      +6.31%*
record_create_1000                                 [   1.5 ms ...   1.6 ms ]      +4.81%*
record_flat_access_1                               [  12.1 us ...  13.1 us ]      +7.83%*
record_flat_access_10                              [  13.0 us ...  14.2 us ]      +8.92%*
record_flat_access_100                             [  21.1 us ...  22.8 us ]      +7.74%*
record_flat_access_1000                            [  92.2 us ... 102.2 us ]     +10.80%*
record_nested_access_1                             [  12.4 us ...  13.4 us ]      +8.37%*
record_nested_access_2                             [  12.8 us ...  13.7 us ]      +7.03%*
record_nested_access_4                             [  13.8 us ...  14.6 us ]      +6.17%*
record_nested_access_8                             [  15.1 us ...  16.4 us ]      +8.55%*
record_nested_access_16                            [  18.1 us ...  20.0 us ]     +10.37%*
record_nested_access_32                            [  24.4 us ...  26.7 us ]      +9.16%*
record_nested_access_64                            [  36.7 us ...  40.2 us ]      +9.53%*
record_nested_access_128                           [  60.6 us ...  66.4 us ]      +9.59%*
table_create_1                                     [  18.1 us ...  19.8 us ]      +9.84%*
table_create_10                                    [  33.5 us ...  37.9 us ]     +13.23%*
table_create_100                                   [ 184.4 us ... 211.9 us ]     +14.96%*
table_create_1000                                  [   1.8 ms ...   2.0 ms ]     +13.10%*
table_get_1                                        [  16.7 us ...  17.9 us ]      +6.93%*
table_get_10                                       [  18.2 us ...  19.4 us ]      +6.28%*
table_get_100                                      [  33.4 us ...  34.6 us ]      +3.79%*
table_get_1000                                     [ 176.5 us ... 178.3 us ]      +1.07%*
table_select_1                                     [  15.6 us ...  17.1 us ]     +10.04%*
table_select_10                                    [  20.9 us ...  22.5 us ]      +7.29%*
table_select_100                                   [  66.8 us ...  68.3 us ]      +2.31%*
table_select_1000                                  [ 522.8 us ... 523.8 us ]      +0.20%
eval_interleave_100                                [ 387.9 us ... 396.6 us ]      +2.24%*
eval_interleave_1000                               [   3.3 ms ...   3.3 ms ]      -0.43%
eval_interleave_10000                              [  32.2 ms ...  31.8 ms ]      -1.27%
eval_interleave_with_interrupt_100                 [ 388.4 us ... 396.1 us ]      +1.99%*
eval_interleave_with_interrupt_1000                [   3.3 ms ...   3.2 ms ]      -1.65%*
eval_interleave_with_interrupt_10000               [  31.9 ms ...  31.7 ms ]      -0.82%
eval_for_1                                         [  28.7 us ...  31.1 us ]      +8.30%*
eval_for_10                                        [  29.7 us ...  31.9 us ]      +7.31%*
eval_for_100                                       [  35.5 us ...  37.9 us ]      +6.78%*
eval_for_1000                                      [  95.1 us ...  96.5 us ]      +1.51%*
eval_for_10000                                     [ 688.7 us ... 681.5 us ]      -1.03%*
eval_each_1                                        [  27.1 us ...  28.7 us ]      +5.89%*
eval_each_10                                       [  28.6 us ...  30.2 us ]      +5.84%*
eval_each_100                                      [  41.0 us ...  41.7 us ]      +1.67%*
eval_each_1000                                     [ 166.8 us ... 156.8 us ]      -6.01%*
eval_each_10000                                    [   1.4 ms ...   1.3 ms ]      -8.15%*
eval_par_each_1                                    [  49.1 us ...  53.4 us ]      +8.91%*
eval_par_each_10                                   [  52.1 us ...  54.6 us ]      +4.76%*
eval_par_each_100                                  [  69.4 us ...  73.8 us ]      +6.26%*
eval_par_each_1000                                 [ 224.0 us ... 227.8 us ]      +1.68%*
eval_par_each_10000                                [   1.7 ms ...   1.7 ms ]      +0.62%*
eval_default_config                                [   2.3 ms ...   2.4 ms ]      +4.03%*
eval_default_env                                   [ 422.0 us ... 463.3 us ]      +9.79%*
encode_json_100_5                                  [  84.1 us ...  82.9 us ]      -1.48%*
encode_json_10000_15                               [  24.0 ms ...  23.6 ms ]      -1.46%*
encode_msgpack_100_5                               [  57.3 us ...  57.5 us ]      +0.27%
encode_msgpack_10000_15                            [  15.6 ms ...  15.8 ms ]      +1.09%*
decode_json_100_5                                  [ 207.4 us ... 206.0 us ]      -0.67%*
decode_json_10000_15                               [  58.6 ms ...  58.3 ms ]      -0.56%
decode_msgpack_100_5                               [ 107.3 us ... 108.9 us ]      +1.55%*
decode_msgpack_10000_15                            [  32.7 ms ...  32.6 ms ]      -0.41%
```

File: crates/nu-command/src/system/nu_check.rs [hunk 2]
```diff
@@ -87,7 +87,7 @@ impl Command for NuCheck {
                         &path_str.item,
                         engine_state,
                         stack,
-                        get_dirs_var_from_call(call),
+                        get_dirs_var_from_call(stack, call), // FIXME
```
[the remark this answers is not in the document]
Comment (devyn) [reply crates/nu-command/src/system/nu_check.rs:90]: I think I might have had `get_dirs_var_from_call` do something hacky before but already fixed it. I'll remove that comment

**Pull request merged**